File: VIPClient.cpp
#include "VIPClient.hpp"
#include <iostream>
#include <stdexcept>

// Конструктор с вызовом конструктора базового класса
VIPClient::VIPClient(const std::string& name, const std::string& phone, 
                     const std::string& vipLevel, double discount)
    : Client(name, phone), vipLevel(vipLevel), discount(discount)
{
    // Вызов конструктора базового класса Client через список инициализации
    try {
        if (discount < 0.0 || discount > 100.0) {
            throw std::invalid_argument("Discount must be between 0 and 100");
        }
        // VIP клиенты сразу получают статус Gold
        this->status = "Gold";
    } catch (const std::exception& e) {
        std::cerr << "Error in VIPClient constructor: " << e.what() << std::endl;
    }
}

// Конструктор копирования
VIPClient::VIPClient(const VIPClient& other)
    : Client(other), vipLevel(other.vipLevel), discount(other.discount)
{
    // Вызов конструктора копирования базового класса
}

void VIPClient::setVipLevel(const std::string& level) {
    this->vipLevel = level;
}

// Переопределение метода базового класса с вызовом базового метода
void VIPClient::showInfo() const {
    Client::showInfo(); // Вызов метода базового класса Client
    std::cout << "  VIP Level: " << this->vipLevel << std::endl;
    std::cout << "  Discount: " << this->discount << "%" << std::endl;
    std::cout << "  Discounted Total: " << this->calculateValue() << std::endl;
}

// Оператор присваивания
VIPClient& VIPClient::operator=(const VIPClient& other) {
    if (this != &other) {
        // Вызов оператора присваивания базового класса
        Client::operator=(other);
        this->vipLevel = other.vipLevel;
        this->discount = other.discount;
    }
    return *this;
}

VIPClient& VIPClient::operator=(const Client& other) {
    Client::operator=(other); // Присваивание только базовой части
    // Поля, специфичные для VIPClient, остаются неизменными
    return *this;
}

VIPClient& VIPClient::operator=(const RestaurantEntity& other) {
    RestaurantEntity::operator=(other); // Присваивание только базовой части
    // Поля, специфичные для VIPClient, остаются неизменными
    return *this;
}

File: Java/VIPClient.java
// Производный класс VIPClient наследуется от Client и реализует множественное наследование
// (от класса Client, который наследуется от RestaurantEntity, и интерфейсов Payable, Displayable)
public class VIPClient extends Client {
    private String vipLevel;
    private double discount;

    public VIPClient(String name, String phone, String vipLevel, double discount) {
        super(name, phone); // Вызов конструктора базового класса Client
        this.vipLevel = vipLevel;
        this.discount = discount;
        // VIP клиенты сразу получают статус Gold
        this.status = "Gold";
    }

    // Переопределение метода базового класса с вызовом базового метода
    @Override
    public void showInfo() {
        super.showInfo(); // Вызов метода базового класса Client
        System.out.println("  VIP Level: " + vipLevel);
        System.out.println("  Discount: " + discount + "%");
        System.out.println("  Discounted Total: " + calculateValue());
    }

    @Override
    public String getType() {
        return "VIPClient";
    }

    @Override
    public double calculateValue() {
        return totalSpent * (1.0 - discount / 100.0);
    }

    // Переопределение метода интерфейса Payable
    @Override
    public double getAmount() {
        return calculateValue(); // С учетом скидки
    }

    @Override
    public String getPaymentInfo() {
        return "VIP Client " + getName() + " discounted total: " + calculateValue() + 
               " (original: " + totalSpent + ", discount: " + discount + "%)";
    }

    // Новые методы
    public void setVipLevel(String level) {
        this.vipLevel = level;
    }

    public String getVipLevel() {
        return vipLevel;
    }

    public double getDiscount() {
        return discount;
    }
}

